src/c2ffi/generator.lisp

```diff
@@ -0,0 +1,755 @@
+                                  ;; either a pathname or a string (will be copied as is),
+                                  ;; or a function that will be funcall'd with one argument
+                                  ;; to emit a form (i.e. OUTPUT/CODE).
+                                  prelude
+                                  (output-encoding asdf:*default-encoding*)
+                                  (emit-generated-name-mappings t)
+                                  (include-sources :all)
+                                  exclude-sources
+                                  (include-definitions :all)
+                                  exclude-definitions)
+  "Generates a lisp file with CFFI definitions from C2FFI-SPEC-FILE.
+PACKAGE-NAME will be overwritten, it assumes full control over the
+target package."
+  (check-type c2ffi-spec-file (or pathname string))
+  (macrolet ((@ (var)
+                 `(setf ,var (compile-rules ,var))))
+    (@ include-sources)
+    (@ exclude-sources)
+    (@ include-definitions)
+    (@ exclude-definitions))
+  (with-standard-io-syntax
+    (with-input-from-file (in c2ffi-spec-file :external-format (asdf/driver:encoding-external-format :utf-8))
+      (with-output-to-file (*c2ffi-output-stream* output :if-exists :supersede
+                            :external-format (asdf/driver:encoding-external-format output-encoding))
+        (let* ((*package* (or (find-package package-name)
+                              (make-package package-name)))
+               ;; Make sure we use an uninterned symbol, so that it's neutral to READTABLE-CASE.
+               (package-name (make-symbol (package-name *package*)))
+               ;; Let's rebind a copy, so that when we are done with the generation
+               ;; then the CFFI type repository is also reverted back to the
+               ;; previous state. This avoids future redefinition warning if they
+               ;; get added to defctype and friends.
+               (cffi::*type-parsers* (copy-hash-table cffi::*type-parsers*))
+               (*anon-name-counter* 0)
+               (*anon-entities* (make-hash-table))
+               (*generated-names* (mapcar (lambda (key)
+                                            `(,key . ,(make-hash-table :test 'equal)))
+                                          +name-kinds+))
+               (*allow-pointer-type-simplification* allow-pointer-type-simplification)
+               (*allow-skipping-struct-fields* allow-skipping-struct-fields)
+               (*assume-struct-by-value-support* assume-struct-by-value-support)
+               (*ffi-name-transformer* (canonicalize-transformer-hook ffi-name-transformer))
+               (*ffi-type-transformer* (canonicalize-transformer-hook ffi-type-transformer))
+               (json (json:decode-json in)))
+          (output/string +generated-file-header+)
+          ;; some forms that are always emitted
+          (mapcar 'output/code `((defpackage ,package-name)
```

Comment:
unfortunately defpackage's behavior is undefined when the name already names an existing package. (will do the suggestions anyway, better than nothing)